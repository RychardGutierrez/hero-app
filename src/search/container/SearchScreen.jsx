import { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { parse } from 'query-string';

import { useForm } from '../../shared/hooks/useForm';
import { handleSearch } from '../../UI/tools/handles';
import { getHeroByName } from './../../shared/selectors';

import HeroCard from './../../hero/components/HeroCard';

const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = parse(location.search);

  const { values, handleInputChange } = useForm({ searchText: q });

  const { searchText } = values;

  const { data, successfull } = useMemo(
    () => handleSearch(q, getHeroByName),
    [q],
  );

  const submitHero = (e, searchText) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };
  return (
    <div className="contaier mt-5">
      <h1 className="text-center mb-5">Search screen</h1>
      <div className="row">
        <div className="col-5">
          <h2>Search</h2>
          <hr />
          <form onSubmit={(e) => submitHero(e, searchText)}>
            <input
              type="text"
              placeholder="Search a Hero"
              className="form-control"
              autoComplete="off"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
            />
            <button className="btn btn-outline-primary mt-1" type="submit">
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          {!!successfull ? (
            <>
              <h2 className="text-center">Hero Result - ({data.length})</h2>
              <hr />
              <div className="animate__animated animate__fadeIn">
                {data.map((hero) => (
                  <HeroCard key={hero.id} {...hero} />
                ))}
              </div>
            </>
          ) : !q.length ? (
            <div className="mt-5">
              <hr />
              <div className="alert alert-info">Hero found...</div>
            </div>
          ) : (
            <div className="mt-5">
              <hr />
              <div className="alert alert-danger">Hero do not found...</div>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
