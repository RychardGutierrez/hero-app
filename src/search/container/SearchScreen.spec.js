import { mount } from 'enzyme';
import SearchScreen from './SearchScreen';
import { MemoryRouter } from 'react-router-dom';

const mockNativate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNativate,
}));

describe('SearchScreen', () => {
  it('Should render component defult', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>,
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('.alert.alert-info').text().trim()).toBe(
      'Hero found...',
    );
  });

  it('Should render Batman and input with the value of the queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>,
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('input').prop('value')).toBe('batman');
  });

  it('Should render Hero do not found and input with the value of the queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>,
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('.alert.alert-danger').text().trim()).toBe(
      'Hero do not found...',
    );
  });

  it('Should call navigate to change screen', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>,
    );

    wrapper.find('input').simulate('change', {
      target: { name: 'searchText', value: 'batman' },
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {},
    });

    expect(mockNativate).toHaveBeenCalledWith('?q=batman');
  });
});
