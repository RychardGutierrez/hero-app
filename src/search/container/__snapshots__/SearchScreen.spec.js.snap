// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen Should render Batman and input with the value of the queryString 1`] = `
Array [
  <h1
    className="text-center"
  >
     Search screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h1
        className="text-center"
      >
        Hero Result - (
        1
        )
      </h1>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
      >
        <div
          className="col"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img"
                  src="/assets/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    More..
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`SearchScreen Should render Hero do not found and input with the value of the queryString 1`] = `
Array [
  <h1
    className="text-center"
  >
     Search screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <hr />
      <div
        className="alert alert-danger"
      >
        Hero do not found...
      </div>
    </div>
  </div>,
]
`;

exports[`SearchScreen Should render component defult 1`] = `
Array [
  <h1
    className="text-center"
  >
     Search screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <hr />
      <div
        className="alert alert-info"
      >
        Hero found...
      </div>
    </div>
  </div>,
]
`;
